cpp/include/raft/core/managed_mdarray.hpp

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <raft/core/managed_container_policy.hpp>
+#include <raft/core/managed_mdspan.hpp>
+#include <raft/core/mdarray.hpp>
+#include <raft/core/resources.hpp>
+
+namespace raft {
+
+/**
+ * @brief mdarray with managed container policy
+ * @tparam ElementType the data type of the elements
+ * @tparam Extents defines the shape
+ * @tparam LayoutPolicy policy for indexing strides and layout ordering
+ * @tparam ContainerPolicy storage and accessor policy
+ */
+template <typename ElementType,
+          typename Extents,
+          typename LayoutPolicy    = layout_c_contiguous,
+          typename ContainerPolicy = managed_uvector_policy<ElementType>>
+using managed_mdarray =
+  mdarray<ElementType, Extents, LayoutPolicy, managed_accessor<ContainerPolicy>>;
+
+/**
+ * @brief Shorthand for 0-dim host mdarray (scalar).
+ * @tparam ElementType the data type of the scalar element
+ * @tparam IndexType the index type of the extents
+ */
+template <typename ElementType, typename IndexType = std::uint32_t>
+using managed_scalar = managed_mdarray<ElementType, scalar_extent<IndexType>>;
+
+/**
+ * @brief Shorthand for 1-dim managed mdarray.
+ * @tparam ElementType the data type of the vector elements
+ * @tparam IndexType the index type of the extents
+ * @tparam LayoutPolicy policy for strides and layout ordering
+ */
+template <typename ElementType,
+          typename IndexType    = std::uint32_t,
+          typename LayoutPolicy = layout_c_contiguous>
+using managed_vector = managed_mdarray<ElementType, vector_extent<IndexType>, LayoutPolicy>;
+
+/**
+ * @brief Shorthand for c-contiguous managed matrix.
+ * @tparam ElementType the data type of the matrix elements
+ * @tparam IndexType the index type of the extents
+ * @tparam LayoutPolicy policy for strides and layout ordering
+ */
+template <typename ElementType,
+          typename IndexType    = std::uint32_t,
+          typename LayoutPolicy = layout_c_contiguous>
+using managed_matrix = managed_mdarray<ElementType, matrix_extent<IndexType>, LayoutPolicy>;
+
+/**
+ * @brief Create a managed mdarray.
+ * @tparam ElementType the data type of the matrix elements
+ * @tparam IndexType the index type of the extents
+ * @tparam LayoutPolicy policy for strides and layout ordering
+ * @param handle raft::resources
+ * @param exts dimensionality of the array (series of integers)
+ * @return raft::managed_mdarray
+ */
+template <typename ElementType,
+          typename IndexType    = std::uint32_t,
+          typename LayoutPolicy = layout_c_contiguous,
+          size_t... Extents>
+auto make_managed_mdarray(raft::resources const& handle, extents<IndexType, Extents...> exts)
+{
+  using mdarray_t = managed_mdarray<ElementType, decltype(exts), LayoutPolicy>;
+
+  typename mdarray_t::mapping_type layout{exts};
+  typename mdarray_t::container_policy_type policy{};
+
+  return mdarray_t{handle, layout, policy};
+}
+
```

Comment:
I don't think we had this prior to this PR, and if I previously introduced that overload, it was probably a mistake. There's no way to enforce that an `rmm` device memory resource actually returns managed memory, so it would get fairly dangerous to label an `mdarray` as being `managed` in that case. I think the best options for downstream users would indeed be the alternative you mentioned and then explicitly wrap it in an mdspan with a managed accessor _or_ to create their own container policy that definitely provides managed memory. Without a better type-based enforcement for a memory resource to make managed allocations, I think that's as far as we should go for now.

Reply:
Yes, sounds good!